x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeyGenerator.java

```diff
@@ -56,4 +58,25 @@ public void generateApiKey(Authentication authentication, CreateApiKeyRequest re
 
     }
 
+    public void generateApiKeyForEnrollment(Authentication authentication, CreateEnrollmentTokenRequest request,
+                                                    ActionListener<CreateApiKeyResponse> listener) {
+        if (authentication == null) {
+            listener.onFailure(new ElasticsearchSecurityException("no authentication available to generate API key for enrollment token"));
+        }
+        if (apiKeyService.isInEnrollmentMode () == false) {
+            listener.onFailure(new IllegalArgumentException("Enrollment mode is not enabled"));
+        }
+        rolesStore.getRoleDescriptors(new HashSet<>(Collections.singleton("enroll")),
```

Comment:
getRoleDescriptors() first argument is a role name, not a privilege. When this moves to the transport action, you need to create the limiting role descriptor of the API key yourself and then add that to the newly created CreateApiKeyRequest